File: examples/agent-connector.py
"""
Ximo Mall Agent 连接器 (Python 版)

用于外部 Python 应用调用 Ximo Mall AI Agent。

安装依赖：
    pip install httpx

使用方式：
    agent = XimoMallAgent("http://localhost:3000")
    response = agent.chat("帮我生成红烧牛肉面详情页")
    print(response.text)
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import AsyncIterator, Callable, Optional

import httpx


# ---------------------------------------------------------------------------
# 类型定义
# ---------------------------------------------------------------------------

@dataclass
class ToolCall:
    tool_name: str
    tool_call_id: str
    args: dict


@dataclass
class ToolResult:
    tool_name: str
    tool_call_id: str
    result: object


@dataclass
class AgentResponse:
    """Agent 完整响应"""
    text: str = ""
    reasoning: str = ""
    tool_calls: list[ToolCall] = field(default_factory=list)
    tool_results: list[ToolResult] = field(default_factory=list)
    finish_reason: Optional[str] = None
    usage: Optional[dict] = None


@dataclass
class ChatOptions:
    """对话选项"""
    thread_id: Optional[str] = None
    resource_id: Optional[str] = None
    images: Optional[list[str]] = None
    on_text_delta: Optional[Callable[[str], None]] = None
    on_reasoning_delta: Optional[Callable[[str], None]] = None
    on_tool_call: Optional[Callable[[ToolCall], None]] = None
    on_tool_result: Optional[Callable[[ToolResult], None]] = None


# ---------------------------------------------------------------------------
# 连接器
# ---------------------------------------------------------------------------

class XimoMallAgent:
    """Ximo Mall AI Agent 客户端"""

    def __init__(self, base_url: str = "http://localhost:3000"):
        self.base_url = base_url.rstrip("/")

    def health(self) -> dict:
        """检查连接状态，status === 'ok' 表示正常"""
        import asyncio
        return asyncio.run(self.health_async())

    async def health_async(self) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(f"{self.base_url}/api/agent/chat")
            resp.raise_for_status()
            return resp.json()

    def tools(self) -> dict:
        """获取 Agent 可调用工具列表"""
        import asyncio
        return asyncio.run(self.tools_async())

    async def tools_async(self) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(f"{self.base_url}/api/agent/info")
            resp.raise_for_status()
            payload = resp.json()
            return payload.get("data", payload)

    def chat(
        self,
        message: str,
        history: list[dict] | None = None,
        options: ChatOptions | None = None,
    ) -> AgentResponse:
        """同步发送对话消息"""
        import asyncio
        return asyncio.run(self.chat_async(message, history, options))

    async def chat_async(
        self,
        message: str,
        history: list[dict] | None = None,
        options: ChatOptions | None = None,
    ) -> AgentResponse:
        """异步发送对话消息"""
        options = options or ChatOptions()
        messages = list(history or [])
        messages.append({"role": "user", "content": message})

        body: dict = {"messages": messages}
        if options.thread_id:
            body["threadId"] = options.thread_id
        if options.resource_id:
            body["resourceId"] = options.resource_id
        if options.images:
            body["images"] = options.images

        async with httpx.AsyncClient(timeout=300) as client:
            async with client.stream(
                "POST",
                f"{self.base_url}/api/agent/chat",
                json=body,
                headers={"Content-Type": "application/json"},
            ) as response:
                response.raise_for_status()
                return await self._parse_sse_stream(response, options)

    # ------------------------------------------------------------------
    # SSE 解析
    # ------------------------------------------------------------------

    async def _parse_sse_stream(
        self,
        response: httpx.Response,
        options: ChatOptions,
    ) -> AgentResponse:
        result = AgentResponse()
        buffer = ""

        async for chunk in response.aiter_text():
            buffer += chunk
            parts = buffer.split("\n\n")
            buffer = parts.pop()  # 最后一段可能不完整

            for part in parts:
                event = self._parse_sse_event(part)
                if event is None:
                    continue

                event_type, data = event

                if event_type == "text":
                    delta = data.get("delta", "")
                    result.text += delta
                    if options.on_text_delta:
                        options.on_text_delta(delta)

                elif event_type == "reasoning":
                    delta = data.get("delta", "")
                    result.reasoning += delta
                    if options.on_reasoning_delta:
                        options.on_reasoning_delta(delta)

                elif event_type == "tool_call":
                    tc = ToolCall(
                        tool_name=data["toolName"],
                        tool_call_id=data["toolCallId"],
                        args=data.get("args", {}),
                    )
                    result.tool_calls.append(tc)
                    if options.on_tool_call:
                        options.on_tool_call(tc)

                elif event_type == "tool_result":
                    tr = ToolResult(
                        tool_name=data["toolName"],
                        tool_call_id=data["toolCallId"],
                        result=data.get("result"),
                    )
                    result.tool_results.append(tr)
                    if options.on_tool_result:
                        options.on_tool_result(tr)

                elif event_type == "done":
                    result.finish_reason = data.get("finishReason", "stop")
                    result.usage = data.get("usage")

                elif event_type == "error":
                    raise RuntimeError(data.get("message", "Agent 执行失败"))

        return result

    @staticmethod
    def _parse_sse_event(raw: str) -> tuple[str, dict] | None:
        event_type = "message"
        data_str = ""

        for line in raw.split("\n"):
            if line.startswith("event: "):
                event_type = line[7:].strip()
            elif line.startswith("data: "):
                data_str = line[6:]

        if not data_str:
            return None

        try:
            return event_type, json.loads(data_str)
        except json.JSONDecodeError:
            return None

    # ------------------------------------------------------------------
    # 图片工具
    # ------------------------------------------------------------------

    @staticmethod
    def image_file_to_data_url(file_path: str) -> str:
        """读取本地图片文件并转为 data URL"""
        path = Path(file_path)
        mime_map = {
            ".png": "image/png",
            ".jpg": "image/jpeg",
            ".jpeg": "image/jpeg",
            ".webp": "image/webp",
            ".gif": "image/gif",
        }
        mime_type = mime_map.get(path.suffix.lower(), "image/png")
        b64 = base64.b64encode(path.read_bytes()).decode()
        return f"data:{mime_type};base64,{b64}"


# ---------------------------------------------------------------------------
# 使用示例
# ---------------------------------------------------------------------------

async def main():
    agent = XimoMallAgent("http://localhost:3000")

    # 示例 1：简单对话
    response = await agent.chat_async("帮我生成红烧牛肉面详情页")
    print(f"回复: {response.text}")
    print(f"思考: {response.reasoning[:200]}...")
    print(f"工具调用: {len(response.tool_calls)} 次")

    # 示例 2：带图片的对话
    # image_url = XimoMallAgent.image_file_to_data_url("./product.jpg")
    # response = await agent.chat_async(
    #     "分析这张产品图",
    #     options=ChatOptions(images=[image_url]),
    # )

    # 示例 3：多轮对话（使用 threadId 保持记忆）
    # thread_id = "my-session-001"
    # r1 = await agent.chat_async(
    #     "帮我生成红烧牛肉面详情页",
    #     options=ChatOptions(thread_id=thread_id),
    # )
    # r2 = await agent.chat_async(
    #     "换一种风格",
    #     history=[{"role": "user", "content": "帮我生成红烧牛肉面详情页"}, {"role": "assistant", "content": r1.text}],
    #     options=ChatOptions(thread_id=thread_id),
    # )

    # 示例 4：实时流式回调
    # def on_text(delta: str):
    #     print(delta, end="", flush=True)
    #
    # response = await agent.chat_async(
    #     "生成详情页",
    #     options=ChatOptions(on_text_delta=on_text),
    # )


if __name__ == "__main__":
    import asyncio
    asyncio.run(main())

File: examples/agent-connector.ts
/**
 * Ximo Mall Agent 连接器
 *
 * 用于外部应用调用 Ximo Mall AI Agent 的客户端封装。
 * 支持对话式交互、图片上传、工具调用追踪、思考过程展示。
 *
 * 使用方式：
 *   const agent = new XimoMallAgent("http://localhost:3000");
 *   const response = await agent.chat("帮我生成红烧牛肉面详情页");
 *   console.log(response.text);
 */

// ---------------------------------------------------------------------------
// 类型定义
// ---------------------------------------------------------------------------

export interface AgentMessage {
  role: "user" | "assistant";
  content: string;
}

export interface ToolCall {
  toolName: string;
  toolCallId: string;
  args: Record<string, unknown>;
}

export interface ToolResult {
  toolName: string;
  toolCallId: string;
  result: unknown;
}

export interface AgentResponse {
  /** Agent 回复的完整文本 */
  text: string;
  /** Agent 的思考过程 */
  reasoning: string;
  /** 工具调用记录 */
  toolCalls: ToolCall[];
  /** 工具返回结果 */
  toolResults: ToolResult[];
  /** 结束原因 */
  finishReason: string | null;
  /** Token 用量 */
  usage: Record<string, number> | null;
}

export interface ChatOptions {
  /** 会话线程 ID，用于记忆隔离（多轮对话时传入同一值） */
  threadId?: string;
  /** 资源 ID */
  resourceId?: string;
  /** 图片 data URL 数组（仅关联到最后一条用户消息） */
  images?: string[];
  /** 实时回调：收到文本片段 */
  onTextDelta?: (delta: string) => void;
  /** 实时回调：收到思考片段 */
  onReasoningDelta?: (delta: string) => void;
  /** 实时回调：工具被调用 */
  onToolCall?: (call: ToolCall) => void;
  /** 实时回调：工具返回结果 */
  onToolResult?: (result: ToolResult) => void;
}

// ---------------------------------------------------------------------------
// 连接器
// ---------------------------------------------------------------------------

export class XimoMallAgent {
  private readonly baseUrl: string;

  /**
   * @param baseUrl Ximo Mall 服务地址，默认 http://localhost:3000
   */
  constructor(baseUrl: string = "http://localhost:3000") {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  /**
   * 检查连接状态
   *
   * @returns 服务状态信息，status === "ok" 表示连接正常
   */
  async health(): Promise<{ status: string; timestamp: string }> {
    const response = await fetch(`${this.baseUrl}/api/agent/chat`, {
      method: "GET",
    });
    if (!response.ok) {
      throw new Error(`Agent 服务不可用 (${response.status})`);
    }
    return response.json();
  }

  /**
   * 获取 Agent 可调用工具列表
   *
   * @returns Agent 信息和工具列表
   */
  async tools(): Promise<{
    agent: { id: string; name: string };
    tools: Array<{ id: string; description: string }>;
  }> {
    const response = await fetch(`${this.baseUrl}/api/agent/info`, {
      method: "GET",
    });
    if (!response.ok) {
      throw new Error(`获取工具列表失败 (${response.status})`);
    }
    const payload = await response.json();
    return payload.data ?? payload;
  }

  /**
   * 发送对话消息
   *
   * @param message 用户消息文本
   * @param history 历史对话记录（可选，多轮对话时传入）
   * @param options 选项（线程ID、图片、实时回调等）
   * @returns Agent 完整响应
   */
  async chat(
    message: string,
    history: AgentMessage[] = [],
    options: ChatOptions = {},
  ): Promise<AgentResponse> {
    const messages = [...history, { role: "user" as const, content: message }];

    const body: Record<string, unknown> = { messages };
    if (options.threadId) body.threadId = options.threadId;
    if (options.resourceId) body.resourceId = options.resourceId;
    if (options.images && options.images.length > 0) body.images = options.images;

    const response = await fetch(`${this.baseUrl}/api/agent/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Agent API 错误 (${response.status}): ${text}`);
    }

    return this.parseSSEStream(response, options);
  }

  /**
   * 上传本地图片文件并转为 data URL
   */
  static fileToDataUrl(
    buffer: Buffer,
    mimeType: string = "image/png",
  ): string {
    return `data:${mimeType};base64,${buffer.toString("base64")}`;
  }

  /**
   * 从文件路径读取图片并转为 data URL（Node.js 环境）
   */
  static async imageFileToDataUrl(
    filePath: string,
  ): Promise<string> {
    const fs = await import("fs");
    const path = await import("path");
    const buffer = fs.readFileSync(filePath);
    const ext = path.extname(filePath).toLowerCase();
    const mimeMap: Record<string, string> = {
      ".png": "image/png",
      ".jpg": "image/jpeg",
      ".jpeg": "image/jpeg",
      ".webp": "image/webp",
      ".gif": "image/gif",
    };
    const mimeType = mimeMap[ext] ?? "image/png";
    return XimoMallAgent.fileToDataUrl(buffer, mimeType);
  }

  // -------------------------------------------------------------------------
  // SSE 解析
  // -------------------------------------------------------------------------

  private async parseSSEStream(
    response: Response,
    options: ChatOptions,
  ): Promise<AgentResponse> {
    const result: AgentResponse = {
      text: "",
      reasoning: "",
      toolCalls: [],
      toolResults: [],
      finishReason: null,
      usage: null,
    };

    const reader = response.body!.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // 按双换行分割 SSE 事件
      const parts = buffer.split("\n\n");
      buffer = parts.pop()!; // 最后一段可能不完整，保留

      for (const part of parts) {
        const event = this.parseSSEEvent(part);
        if (!event) continue;

        switch (event.type) {
          case "text": {
            const delta = event.data.delta as string;
            result.text += delta;
            options.onTextDelta?.(delta);
            break;
          }
          case "reasoning": {
            const delta = event.data.delta as string;
            result.reasoning += delta;
            options.onReasoningDelta?.(delta);
            break;
          }
          case "tool_call": {
            const call: ToolCall = {
              toolName: event.data.toolName as string,
              toolCallId: event.data.toolCallId as string,
              args: event.data.args as Record<string, unknown>,
            };
            result.toolCalls.push(call);
            options.onToolCall?.(call);
            break;
          }
          case "tool_result": {
            const tr: ToolResult = {
              toolName: event.data.toolName as string,
              toolCallId: event.data.toolCallId as string,
              result: event.data.result,
            };
            result.toolResults.push(tr);
            options.onToolResult?.(tr);
            break;
          }
          case "done": {
            result.finishReason = (event.data.finishReason as string) ?? "stop";
            result.usage = (event.data.usage as Record<string, number>) ?? null;
            break;
          }
          case "error": {
            throw new Error((event.data.message as string) ?? "Agent 执行失败");
          }
        }
      }
    }

    return result;
  }

  private parseSSEEvent(
    raw: string,
  ): { type: string; data: Record<string, unknown> } | null {
    let type = "message";
    let dataStr = "";

    for (const line of raw.split("\n")) {
      if (line.startsWith("event: ")) {
        type = line.slice(7).trim();
      } else if (line.startsWith("data: ")) {
        dataStr = line.slice(6);
      }
    }

    if (!dataStr) return null;

    try {
      return { type, data: JSON.parse(dataStr) };
    } catch {
      return null;
    }
  }
}
